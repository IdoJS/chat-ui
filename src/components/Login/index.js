import React from 'react';
import PropTypes from 'prop-types';


class Login extends React.PureComponent {
  state = {
    avatarClass: '',
    userName: ''
  };

  constructor(props) {
    super(props);

    this.submit = this.submit.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onAvatarChange = this.onAvatarChange.bind(this)
  }

  render() {
  const radios =  [...Array(5).keys()].map((index) => (<li key={index} className='ui image label'>
    <input type='radio' className='radio' name='avatar' id={`avatar_img0${index+1}`}/>
    <div className={`avatar avatar_img0${index+1}`}></div>
  </li>));

    return <form className='ui form' key={0}>
      <h1 className='ui large header'>Login</h1>
      <div class='field'>
        <label>Please enter userName</label>
        <div class='ui left icon input'>
          <input type='text'
                 className='ui input'
                 name='userName'
                 placeholder='Type name here...'
                 value={this.state.userName}
                 onChange={this.onTextChange}/>
          <i class='user circle icon'></i>
        </div>
      </div>

      <div class='field'>
        <label>Please choose avatar</label>
        <div className='ui list'>
          <ul className='form_radio_group' onChange={this.onAvatarChange.bind(this)}>
            {radios}
          </ul>
        </div>
      </div>

      <button className='ui primary button' onClick={this.submit}>Submit</button>
    </form>
  }

  onAvatarChange(ev) {
    this.setState({
      avatarClass: ev.target.id
    })
  }

  onTextChange(ev) {
    this.setState({
      [ev.target.name]: ev.target.value
    });
  }

  submit(ev) {
    ev.preventDefault();
    this.props.onUserCreate({
      userName: this.state.userName,
      avatarClass: this.state.avatarClass
    })
  }
};


Login.propTypes = {
  onUserCreate : PropTypes.func.isRequired
};

Login.defaultProps = {
  onUserCreate : () => {
    console.warn('Please implement onUserCreate');
  }
};

export default Login;