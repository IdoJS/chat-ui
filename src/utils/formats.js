const REGEX_SPLIT_DATE = /[T.]+/;
const LOCAL_OFFSET = new Date().getTimezoneOffset() * 60000;

/**
 * Return split timestamp by timezone
 * [0] - date
 * [1] - time
 * @param timestamp
 */
export const splitTimestamp = (timestamp) => {
  const localTime = timestamp - LOCAL_OFFSET;
  let formattedDateArr;

  try {
    formattedDateArr = new Date(localTime).toISOString().split(REGEX_SPLIT_DATE);
  } catch(e){
    formattedDateArr = ['','',''];
  }
  return formattedDateArr;
};