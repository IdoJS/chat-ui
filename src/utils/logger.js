export const logPrint = (...data) => {
  if (!process.env.NODE_ENV || process.env.NODE_ENV === 'development') {
    console.log('LOGGER ', data);
  }
};