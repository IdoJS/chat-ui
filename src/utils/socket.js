import io from 'socket.io-client';
import mockIo from '../../src/__mocks__/mockSocket';

export const getIO = () => {
  if (process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'dev') {
    return mockIo.connect();
  } else {
    return io;
  }
};