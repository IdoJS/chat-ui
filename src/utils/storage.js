/**
 * Choose storage by ENV
 * @type {Storage}
 */
const storage = !process.env.NODE_ENV || process.env.NODE_ENV === 'development' ? sessionStorage : localStorage;

/**
 * Generate unique ID
 * https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
 * @returns {string}
 */
const guid = () => {
  const s4 = () => {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  };
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
};

const STORAGE_AVATAR_CLASS = 'avatarClass';
const STORAGE_USER_NAME = 'userName';
const STORAGE_USER_ID = 'UserID';

export const getAvatarClass = () => {
  return storage.getItem(STORAGE_AVATAR_CLASS);
};

export const getUserName = () => {
  return storage.getItem(STORAGE_USER_NAME);
};

export const setAvatarClass = (avatar) => {
  storage.setItem(STORAGE_AVATAR_CLASS, avatar);
};

export const setUserName = (userName) => {
  storage.setItem(STORAGE_USER_NAME, userName);
};

export const getUserId = () => {
  return storage.getItem(STORAGE_USER_ID);
};

export const setUserId = () => {
  storage.setItem(STORAGE_USER_ID, guid());
};