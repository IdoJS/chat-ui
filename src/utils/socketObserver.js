import {getIO} from './socket';

const SOCKET_URL = 'https://spotim-demo-chat-server.herokuapp.com';
const SOCKET_EVENT = 'spotim/chat';
const SOCKET_IDENTIFIER = 'socket_identifier';
/**
 * Singleton Observer that listen to socket
 */

let observers = [];

class SocketObserver {

  static instance;

  constructor() {
    const io = getIO();

    if (this.instance) {
      return this.instance;
    }

    this.instance = this;

    window.socket = io(SOCKET_URL);

    window.socket.on(SOCKET_EVENT, (response) => {
      if (response.id === SOCKET_IDENTIFIER) {
        observers.forEach(obs => typeof obs.o === 'function' && obs.o(response));
      }
    });
  }

  /**
   * Add observer
   * @param identifier
   * @param o
   * @returns {SocketObserver}
   */
  subscribe({identifier, o}) {
    observers.push({identifier, o});
    return this;
  }

  /**
   * Remove observer
   * @param identifier
   * @returns {SocketObserver}
   */
  unSubscribe({identifier}) {
    observers = observers.filter(ob => ob.identifier !== identifier);
    return this;
  }

  /**
   * Make socket request
   * @param requestData
   * @returns {SocketObserver}
   */
  send(requestData) {
    window.socket.emit(SOCKET_EVENT, Object.assign(requestData, {id: SOCKET_IDENTIFIER}));
    return this;
  }

  /**
   * Remove all observers
   */
  clear() {
    observers = [];
  }

  /**
   * Get observer list
   * @returns {Array}
   */
  getObservers() {
    return [...observers];
  }
}

export default new SocketObserver();