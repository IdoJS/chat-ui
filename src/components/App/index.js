import React from 'react';
import {getUserName, getAvatarClass, setAvatarClass, setUserName} from '../../utils/storage';
import Login from '../Login';
import ChatRoom from '../../containers/ChatRoom';

class App extends React.PureComponent {

  state = {
    userName: getUserName(),
    avatarClass: getAvatarClass()
  };

  constructor(props) {
    super(props);

    this.onUserCreate = this.onUserCreate.bind(this);
  }

  onUserCreate({userName, avatarClass}) {
    setUserName(userName);
    setAvatarClass(avatarClass);
    this.setState({
      userName,
      avatarClass
    });
  }

  render() {
    return !this.state.userName || !this.state.avatarClass ?
      (<Login onUserCreate={this.onUserCreate}/>) :
      (<ChatRoom/>);
  }
}

export default App;
