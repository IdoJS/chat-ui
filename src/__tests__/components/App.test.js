import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import App from '../../../src/components/App';
import Login from '../../../src/components/Login';
import {setUserName, setAvatarClass, getUserName, getAvatarClass} from '../../utils/storage';

describe('App component', () => {

  beforeEach(()=>{
    global.localStorage.clear();
  });


  it('Should render the login view when storage is empty', () =>{
    const wrapper = shallow(<App/>);
    const child = wrapper.find(Login);
    expect(child).toHaveLength(1);


  });

  it('SnapShot rendering Login', () => {
    const wrapper = shallow(<App/>);
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render the ChatRoom view when storage is contain userName & avatarClass', () =>{
    setUserName('mockUser');
    setAvatarClass('mockAvatar');

    const instance = shallow(<App/>).instance();
    expect(instance.state.userName).toBe(getUserName());
    expect(instance.state.avatarClass).toBe(getAvatarClass());
  });
});