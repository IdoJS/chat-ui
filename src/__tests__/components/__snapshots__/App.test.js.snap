// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component SnapShot rendering Login 1`] = `
<form
  className="ui form"
>
  <h1
    className="ui large header"
  >
    Login
  </h1>
  <div
    class="field"
  >
    <label>
      Please enter userName
    </label>
    <div
      class="ui left icon input"
    >
      <input
        className="ui input"
        name="userName"
        onChange={[Function]}
        placeholder="Type name here..."
        type="text"
        value=""
      />
      <i
        class="user circle icon"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label>
      Please choose avatar
    </label>
    <div
      className="ui list"
    >
      <ul
        className="form_radio_group"
        onChange={[Function]}
      >
        <li
          className="ui image label"
        >
          <input
            className="radio"
            id="avatar_img01"
            name="avatar"
            type="radio"
          />
          <div
            className="avatar avatar_img01"
          />
        </li>
        <li
          className="ui image label"
        >
          <input
            className="radio"
            id="avatar_img02"
            name="avatar"
            type="radio"
          />
          <div
            className="avatar avatar_img02"
          />
        </li>
        <li
          className="ui image label"
        >
          <input
            className="radio"
            id="avatar_img03"
            name="avatar"
            type="radio"
          />
          <div
            className="avatar avatar_img03"
          />
        </li>
        <li
          className="ui image label"
        >
          <input
            className="radio"
            id="avatar_img04"
            name="avatar"
            type="radio"
          />
          <div
            className="avatar avatar_img04"
          />
        </li>
        <li
          className="ui image label"
        >
          <input
            className="radio"
            id="avatar_img05"
            name="avatar"
            type="radio"
          />
          <div
            className="avatar avatar_img05"
          />
        </li>
      </ul>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
  >
    Submit
  </button>
</form>
`;
